### PR TITLE
Fix aliases

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix aliases and add a couple of improvements to the expression parser.

### CODE REVIEW

File: src/Parsers/ExpressionListParsers.cpp [hunk 1]
```diff
@@ -786,6 +789,9 @@ template <TokenType separator, TokenType end>
 class BaseLayer : public Layer
```
Comment (SmitaRKulkarni) [inline src/Parsers/ExpressionListParsers.cpp:789]: I think it's a bit misleading that BaseLayer extends from Layer. Maybe the names can be interchanged ?
But I see in other Layers, that we sometimes extend from Layer and sometimes from BaseLayer, not sure what names would be helpful.

File: src/Parsers/ExpressionListParsers.cpp [hunk 2]
```diff
@@ -1090,6 +1096,8 @@ class ArrayElementLayer : public BaseLayer<TokenType::Comma, TokenType::ClosingS
 class CastLayer : public Layer
 {
 public:
+    CastLayer() : Layer(true, true) {}
```
Comment (SmitaRKulkarni) [inline src/Parsers/ExpressionListParsers.cpp:1099]:
```suggestion
    CastLayer() : Layer(/*allow_alias*/ true, /*allow_alias_without_as_keyword*/ true) {}
```
also in other constructors

File: src/Parsers/ExpressionListParsers.cpp [hunk 3]
```diff
@@ -2427,11 +2442,15 @@ Action ParserExpressionImpl::tryParseOperator(Layers & layers, IParser::Pos & po
 
     if (cur_op == operators_table.end())
     {
+        ParserAlias alias_parser(layers.back()->allow_alias_without_as_keyword);
         auto old_pos = pos;
-        if (layers.back()->allow_alias && ParserAlias(layers.back()->allow_alias_without_as_keyword).parse(pos, tmp, expected))
+        if (!layers.back()->parsed_alias && layers.back()->allow_alias && alias_parser.parse(pos, tmp, expected))
         {
             if (layers.back()->insertAlias(tmp))
```
Comment (SmitaRKulkarni) [inline src/Parsers/ExpressionListParsers.cpp:2449]: I think this condition can be combined to the above if condition

File: src/Parsers/ExpressionListParsers.cpp (hunk 1 above)
Comment (evillique) [reply src/Parsers/ExpressionListParsers.cpp:789]: We have classes `IParser` and `IParserBase` where `IParserBase` extends from `IParser`, so it was done here in a similar way. We can maybe change `BaseLayer` to `OrdinaryLayer` or something similar to that, what do you think?

Comment (SmitaRKulkarni) [reply src/Parsers/ExpressionListParsers.cpp:789]: I was not aware of similar patterns, thanks. But I see that almost all Parsers use IParserBase.
In our case we extend from both, I think the difference is the separator & end. How about LayerWithSeparator  or LayerWithCustomSeparator? I am not a big fan of OrdinaryLayer, but I checked in our code we use OrdinaryXXX / SimpleXXX names, so to be consistent these would do too. Let me know your thoughts

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

File: src/Parsers/ExpressionListParsers.cpp (hunk 1 above)
Comment (evillique) [reply src/Parsers/ExpressionListParsers.cpp]: I think for now LayerWithSeparator is ok for now. I just want to deliver this alias fix before the 22.10 release.

**Review by SmitaRKulkarni — APPROVED** [review]

**Comment by evillique** [comment]
Test failures are unrelated.

**Pull request merged**